### PR TITLE
Automatically shorten long links

### LINKED ISSUE #113: Automatically shorten long links
Some users deliberately post extremely long links for attention. If a link exceeds a certain number of characters, automatically shorten it and insert an ellipsis to indicate that its shortened.

### PR DESCRIPTION
**Pull request by Riott**
Continuing the work started in https://github.com/destinygg/chat-gui/pull/117 shortens links over 90 characters to 80 characters with an elipsis.

image

resolves #113

### CODE REVIEW

File: assets/chat/js/formatters/UrlFormatter.js [hunk 1]
```diff
@@ -71,9 +71,19 @@ export default class UrlFormatter {
       const m = decodedUrl.match(self.linkregex);
       if (m) {
         const encodedUrl = self.encodeUrl(m[0]);
+        const maxUrlLength = 90;
+        let urlText = encodedUrl;
+        if (encodedUrl.length > maxUrlLength) {
```
Comment (11k) [inline assets/chat/js/formatters/UrlFormatter.js:76]: Although it makes no difference behaviorally, this logic has to do with adjusting the URL's text, so it makes more sense to compare the length of `urlText`.

```suggestion
        if (urlText.length > maxUrlLength) {
```

File: assets/chat/js/formatters/UrlFormatter.js [hunk 2]
```diff
@@ -71,9 +71,19 @@ export default class UrlFormatter {
...
+          const middleIndex = Math.floor(maxUrlLength / 2);
+          urlText = `${encodedUrl.slice(
+            0,
+            middleIndex - 3
+          )}...${encodedUrl.slice(middleIndex + 4 - (encodedUrl.length % 2))}`;
```
Comment (11k) [inline assets/chat/js/formatters/UrlFormatter.js:81]: Did you leave this in by mistake?
```suggestion
```

**Review by 11k — CHANGES_REQUESTED** [review]

File: assets/chat/js/formatters/UrlFormatter.js (hunk 1 above)
Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js:76]: Sure will sort

File: assets/chat/js/formatters/UrlFormatter.js (hunk 2 above)
Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js:81]: Sorry not sure what you're referring to here? feel free to reach out on discord to me to point out specifics Riott#9810

Comment (11k) [reply assets/chat/js/formatters/UrlFormatter.js:81]: It seems like you took my advice in [this comment](https://github.com/destinygg/chat-gui/pull/117/files#r1093697650) and [this comment](https://github.com/destinygg/chat-gui/pull/117/files#r1093712665), but left in the URL-shortening logic from the old PR as well. Looks like you're creating a shortened URL twice.

Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js:81]: The statement referenced slices the original url to the middle index (accounting for the elipses) adds the elipsis and then slices from after the elipsis to the end accounting for an odd length string. then the line after reduces that further to the first 40 characters each side of the elipsis, keeping the character length the same size either side of the elipsis (at least that was my intent) I very well might just be blind though and have fucked up somewhere 😆

Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js:81]: looking back a week later it seems pretty hard to read I could push the logic into a function or if we don't really care about perfect placement I can just keep the second line

Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js:81]: disregard I think I'm stupid, will sort and leave as evidence of my stupidity

Comment (11k) [reply assets/chat/js/formatters/UrlFormatter.js:81]: Yup, I think with my suggested change everything works the same, more or less.

Comment (Riott) [reply assets/chat/js/formatters/UrlFormatter.js]: Thank you for your patience 🤣

**Pull request merged**